File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();

		String s = br.readLine();
		int t = Integer.parseInt(s);

		while (t-- > 0) {
			Stack<Character> stk = new Stack<>();
			int flag = 0;

			String input = br.readLine();
			for (int i = 0; i < input.length(); i++) {
				char c = input.charAt(i);
				if (c == '(')
					stk.push(c);
				else {
					if (stk.empty()) {
						sb.append("NO\n");
						flag = 1;
						break;
					}
					stk.pop();
				}
			}
			if (flag == 0) {
				if (!stk.empty())
					sb.append("NO\n");
				else
					sb.append("YES\n");
			}
		}
		System.out.println(sb);
	}
}

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <string>
#include <stack>

using namespace std;

int t;

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);

	cin >> t;

	while(t--) {
		string s;
		int fin = 0;
		stack<char> st;
		cin >> s;
		int len = s.length();

		for (int i = 0; i < len; i++) {
			if (s[i] == '(') {
				st.push(s[i]);
			} else { // )
				if (st.empty()) {
					fin = 1;
					break;
				} else {
					st.pop();
				}
			}
		}
		if (fin || !st.empty()) {
			cout << "NO\n";
		} else if (!fin) {
			cout << "YES\n";
		}
	}

	return 0;
}